### ISSUE TITLE
Firewall SSO Category Comparison bug

### ISSUE THREAD

**Issue by sinusoidal-zz** [inverse-inc/packetfence#486]
The code in lib/pf/firewalsso/Fortigate.pm does a comparison of categories in an 'if' statement.  There seem to be two problems in the existing code (tested on 4.5.1, can't find any duplicate bug reports).

This may also apply to PaloAlto.pm and any other implementations also.

1) 
The node category is forced to lower case in the comparison, but the firewall categories are not.  If you have a firewall category with an upper case character, the match will fail, e.g:

Comparison of 'adult-users' with 'Adult-Users' fails.

```
grep { lc $node_info->{'category'} eq $_ } ( $ConfigFirewallSSO{$firewall_conf}->{'categories'})
```

2) 
The comparison code doesn't seem to be correct.  In order to use grep I believe a 'LIST' object should be passed in as an argument, but $ConfigFirewallSSO{$firewall_conf}->{'categories'} isn't a LIST object, it's comma separated text string pulled directly from a config file?

I'm sure there is a better way of doing this as a one liner, but in my mediocre perl, I fixed it in my implementation with the following changes (the old code is left in commented out):

```
###two new lines prior to comparison to generate list from CSV string in lower case
my $category_list = lc $ConfigFirewallSSO{$firewall_conf}->{'categories'};  #forces LC
my @categories = split( /,/, $category_list);

###OLD COMPARISON CODE###  if (defined($node_info) && (ref($node_info) eq 'HASH') && $node_info->{'status'} eq $pf::node::STATUS_REGISTERED &&  grep { lc $node_info->{'category'} eq $_ } ( $ConfigFirewallSSO{$firewall_conf}->{'categories'})) {

##new if statement uses above created list
if (defined($node_info) && (ref($node_info) eq 'HASH') && grep { lc $node_info->{'category'} eq $_ } @categories) {

```

Thoughts?

**Comment by sinusoidal-zz** [comment]
(I should add my code in that example has also removed the 'STATUS_REGISTERED' check from the IF statement, but that's not relevant to this bug report).

**Comment by julsemaan** [comment]
This has been reworked in the most recent version of PacketFence so this will be a won't fix for 4.5.1

The grep is indeed pretty bad, and you are on the right track with your fix, this is similar to the way it was redone in 5.0

If your patch works fine, you can continue using it and when you upgrade to 5+ you will not need to port it
